src/Extensions/PxPayExtension.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace SilverStripe\Omnipay\Extensions;
+
+use SilverStripe\Core\Extension;
+
+/**
+ * Fix request for PxPay.
+ *
+ * To apply this extension use the following configuration.
+ *
+ * SilverStripe\Omnipay\Service\PaymentService:
+ *  extensions:
+ *    - SilverStripe\Omnipay\Extensions\PxPayExtension
+ *
+ * SilverStripe\Omnipay\PaymentGatewayController:
+ *  extensions:
+ *    - SilverStripe\Omnipay\Extensions\PxPayExtension
+ */
+class PxPayExtension extends Extension
+{
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforePurchase(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeAuthorize(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeCompletePurchase(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeCompleteAuthorize(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param HTTPRequest $request The current request
+     * @param string $gateway the gateway name.
+     *
+     * @return Iterable
+     */
+    public function updatePaymentFromRequest($request, $gateway = null)
+    {
+        if ($gateway == 'PaymentExpress_PxPay') {
+            $identifier = $request->param('Identifier');
+
+            if ($identifier && strlen($identifier) == 16) {
+                // identifier in PxPay is limited to 16 characters so can match
+                // on the start as long as we have 16 characters (i.e matching
+                // on `1` would not be secure).
+                //
+                // Identifiers are normally 30 characters.
+                return Payment::get()->filter([
+                    'Identifier:StartsWith' => $identifier,
```

Comment:
There is an error here with a very low probability of collision occurring, but consequences could be severe. Any identifiers where the first 16 characters are identical would match this query, and Omnipay would happily generate out many more identifiers starting with the same 16 characters.

E.g. `1234567890123456abcd...` and `1234567890123456zyxw...` would both be picked up by this filter, and more payments with identifiers `1234567890123456...` could be generated since they're not identical to anything in the database. All of those transactions would result in the very first transaction being re-processed again every time the gateway came back with a response.

Reply:
I think we should separate the identifier used in the URL from the transaction-id that is being passed to Omnipay. The transaction-id defaults to the Identifier at the moment, but could just as well be `"tnx-<id>"` or something along these lines.

Reply:
Ok sweet can alter to use that. Figured it needed to be a secure string value.